### DISCUSSION TITLE
扫码关注一直提示服务故障的问题

### DISCUSSION THREAD

**Discussion by xx10n31y** [w7corp/easywechat#2298] [Q&A] [answered]
发在 Learnku 那边没什么有用的回复, 也尝试过各种返回方式也还是不行, 不得行只能求助

https://learnku.com/laravel/t/63598

**Comment by overtrue** [comment]
https://mp.weixin.qq.com/debug/ 这里测试，把响应截图

**Reply by xx10n31y** [reply]
> https://mp.weixin.qq.com/debug/ 这里测试，把响应截图

![image](https://user-images.githubusercontent.com/57975788/145946044-5393a3a6-62a5-4fdf-8753-12f687497d32.png)

但业务流程没有受到影响, 也能正常拿到用户的信息, 使用的是兼容模式

**Reply by xx10n31y** [reply]
> https://mp.weixin.qq.com/debug/ 这里测试，把响应截图

之前使用的是测试公众号的明文模式, 后面才使用公司的公众号, 为了安全才使用兼容模式, 虽然业务流程没收到影响, 但一扫码, 就会提示提供的服务出现故障

**Comment by overtrue** [comment]
"微信服务器在五秒内收不到响应会断掉连接，并且重新发起请求，总共重试三次"

把这行注释掉再试试？

```php
$this->handleWxEvent($message, $app);
```

**Reply by xx10n31y** [reply]
> 

这是处理扫码等事件的方法, 如果是在公众号内直接发送文字这些, 倒不会出现故障, 但处理事件的时候就会, 但我觉得代码这一块好像没什么问题, 下面是这个方法的代码
```php

/**
     * @throws AppException|InvalidArgumentException
     */
    protected function handleWxEvent($message, $app)
    {
        // 扫码事件
        // 关注事件
        if (
            (
                ($message['Event'] =='SCAN' && $message['EventKey'] == self::LOGIN_SCENE_ID) ||
                // 未关注扫码时的 EventKey 为 qrscene 拼接上自定义的 eventKey
                ($message['Event'] == 'subscribe' && $message['EventKey'] == 'qrscene_'.self::LOGIN_SCENE_ID)
            )
            &&
            (array_key_exists('Ticket', $message) && $message['Ticket'])
        ) {
            try {
                $ticket = $message['Ticket'];
                $userInfo = $app->user->get($message['FromUserName']);
                [$user, $bindMobile] = $this->userLogin($userInfo);
                // 未绑定手机号
                $data = [
                    'token' => '',
                    'bind' => $bindMobile,
                    'scanned' => 1,
                ];
                if ($bindMobile) {
                    // 已绑定手机号, 直接登录
                    // 进行缓存
                    $token = auth('api')->login($user);
                    // 五分钟有效
                    $data['token'] = $token;
                }
                cache()->set(md5($ticket), $data, 300);
                // cache()->set(md5($ticket) . '_bind_user_id', $user->id, 300);
                cache()->set(md5($ticket) . '_user_raw_info', $userInfo, 300);
            } catch (Exception $exception) {
                Log::error($exception->getMessage());
                throw new AppException($exception->getMessage());
            }
        }
    }

```

**Reply by overtrue** [reply]
> > 
> 
> 
> 
> 这是处理扫码等事件的方法, 如果是在公众号内直接发送文字这些, 倒不会出现故障, 但处理事件的时候就会, 但我觉得代码这一块好像没什么问题, 下面是这个方法的代码
> 
> ```php
> 
> 
> 
> /**
> 
>      * @throws AppException|InvalidArgumentException
> 
>      */
> 
>     protected function handleWxEvent($message, $app)
> 
>     {
> 
>         // 扫码事件
> 
>         // 关注事件
> 
>         if (
> 
>             (
> 
>                 ($message['Event'] =='SCAN' && $message['EventKey'] == self::LOGIN_SCENE_ID) ||
> 
>                 // 未关注扫码时的 EventKey 为 qrscene 拼接上自定义的 eventKey
> 
>                 ($message['Event'] == 'subscribe' && $message['EventKey'] == 'qrscene_'.self::LOGIN_SCENE_ID)
> 
>             )
> 
>             &&
> 
>             (array_key_exists('Ticket', $message) && $message['Ticket'])
> 
>         ) {
> 
>             try {
> 
>                 $ticket = $message['Ticket'];
> 
>                 $userInfo = $app->user->get($message['FromUserName']);
> 
>                 [$user, $bindMobile] = $this->userLogin($userInfo);
> 
>                 // 未绑定手机号
> 
>                 $data = [
> 
>                     'token' => '',
> 
>                     'bind' => $bindMobile,
> 
>                     'scanned' => 1,
> 
>                 ];
> 
>                 if ($bindMobile) {
> 
>                     // 已绑定手机号, 直接登录
> 
>                     // 进行缓存
> 
>                     $token = auth('api')->login($user);
> 
>                     // 五分钟有效
> 
>                     $data['token'] = $token;
> 
>                 }
> 
>                 cache()->set(md5($ticket), $data, 300);
> 
>                 // cache()->set(md5($ticket) . '_bind_user_id', $user->id, 300);
> 
>                 cache()->set(md5($ticket) . '_user_raw_info', $userInfo, 300);
> 
>             } catch (Exception $exception) {
> 
>                 Log::error($exception->getMessage());
> 
>                 throw new AppException($exception->getMessage());
> 
>             }
> 
>         }
> 
>     }
> 
> 
> 
> ```
> 
> 

你要是不试的话，这个事儿就只能另请高明了，我的能力帮不了咯。

> 如果自己认为不会出问题的地方都真的不会出问题，这个世界就没有bug这个东西存在了。
> ----- 鲁迅

**Reply by xx10n31y** [reply]
> > > 
> > 
> > 
> > 这是处理扫码等事件的方法, 如果是在公众号内直接发送文字这些, 倒不会出现故障, 但处理事件的时候就会, 但我觉得代码这一块好像没什么问题, 下面是这个方法的代码
> > ```
> > /**
> > 
> >      * @throws AppException|InvalidArgumentException
> > 
> >      */
> > 
> >     protected function handleWxEvent($message, $app)
> > 
> >     {
> > 
> >         // 扫码事件
> > 
> >         // 关注事件
> > 
> >         if (
> > 
> >             (
> > 
> >                 ($message['Event'] =='SCAN' && $message['EventKey'] == self::LOGIN_SCENE_ID) ||
> > 
> >                 // 未关注扫码时的 EventKey 为 qrscene 拼接上自定义的 eventKey
> > 
> >                 ($message['Event'] == 'subscribe' && $message['EventKey'] == 'qrscene_'.self::LOGIN_SCENE_ID)
> > 
> >             )
> > 
> >             &&
> > 
> >             (array_key_exists('Ticket', $message) && $message['Ticket'])
> > 
> >         ) {
> > 
> >             try {
> > 
> >                 $ticket = $message['Ticket'];
> > 
> >                 $userInfo = $app->user->get($message['FromUserName']);
> > 
> >                 [$user, $bindMobile] = $this->userLogin($userInfo);
> > 
> >                 // 未绑定手机号
> > 
> >                 $data = [
> > 
> >                     'token' => '',
> > 
> >                     'bind' => $bindMobile,
> > 
> >                     'scanned' => 1,
> > 
> >                 ];
> > 
> >                 if ($bindMobile) {
> > 
> >                     // 已绑定手机号, 直接登录
> > 
> >                     // 进行缓存
> > 
> >                     $token = auth('api')->login($user);
> > 
> >                     // 五分钟有效
> > 
> >                     $data['token'] = $token;
> > 
> >                 }
> > 
> >                 cache()->set(md5($ticket), $data, 300);
> > 
> >                 // cache()->set(md5($ticket) . '_bind_user_id', $user->id, 300);
> > 
> >                 cache()->set(md5($ticket) . '_user_raw_info', $userInfo, 300);
> > 
> >             } catch (Exception $exception) {
> > 
> >                 Log::error($exception->getMessage());
> > 
> >                 throw new AppException($exception->getMessage());
> > 
> >             }
> > 
> >         }
> > 
> >     }
> > ```
> 
> 你要是不试的话，这个事儿就只能另请高明了，我的能力帮不了咯。
> 
> > 如果自己认为不会出问题的地方都真的不会出问题，这个世界就没有bug这个东西存在了。
> > ----- 鲁迅

嗯, 我把 `$this->handleWxEvent` 这一块直接改用 return 'event';之后, 在 微信debug 那边就是这样子了
![image](https://user-images.githubusercontent.com/57975788/145948258-cd9409b1-fc53-4e6e-8bd2-ad3b1adbbd9e.png)

**Reply by overtrue** [reply]
那么还报异常么

**Reply by xx10n31y** [reply]
> 那么还报异常么

然后我去公众号扫码试了一下, 扫码之后, 公众号回复 `event`, 但紧跟着还是 `该公众号提供的服务出现故障, 请稍后再试`

**Reply by overtrue** [reply]
确定没接入第三方？

**Reply by xx10n31y** [reply]
> 确定没接入第三方？

确定, 目前就 手机号和邮箱 再加 关注公众号 登录

**Reply by xx10n31y** [reply]
> 确定没接入第三方？

如果不行的话,  那还是用明文模式, 用了兼容模式, 它又不报错而且业务流程又没有受到影响, 很难排查

**Comment by xx10n31y** [comment, marked as answer]
我已经开始怀疑是微信自己的问题了,  使用明文模式, 注释掉config里面的aes_key之后, 扫码之后还是提示 出现故障

**Reply by overtrue** [reply]
你把这个服务端关掉再试

**Reply by xx10n31y** [reply]
> 你把这个服务端关掉再试

**服务器配置(未启用)**, 哈哈哈, 果然, 微信自己的问题, 关掉了服务器配置之后, 再扫码, 还是提示提供的服务出现故障

**Reply by overtrue** [reply]
我怀疑你授权了第三方服务

**Reply by xx10n31y** [reply]
> 我怀疑你授权了第三方服务

授权了第三方服务, 具体指的是什么, 公众号本身的授权还是?

**Reply by overtrue** [reply]
“设置与开发” -> "公众号设置" -> “授权管理”

**Reply by xx10n31y** [reply]
> “设置与开发” -> "公众号设置" -> “授权管理”

![image](https://user-images.githubusercontent.com/57975788/145952170-c8c65d48-1aa5-4a68-9a3e-77f1bdefc049.png)
噢, 看来是这个的问题了, 不过这些授权都是老板弄的,  没办法了, 就这样吧.
抱歉了,  打扰你那么多时间, 感谢你的解答
